### PR TITLE
Audiobridge plugin

### PR DESCRIPTION
**Pull request by vogdb**

### CODE REVIEW

**Comment by vogdb** [comment]
@zazabe please review

**Comment by vogdb** [comment]
@zazabe what about this PR ?

File: src/webrtc/plugin/audiobridge-plugin.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    var jsep = response['jsep'];
+    if (jsep) {
+      this.setRemoteSDP(jsep);
+      return jsep;
+    }
+    return Promise.reject(new Error('Failed sendOffer'));
+  }.bind(this));
+
+  this.addTransaction(transaction);
+  var message = {janus: 'message', body: {request: 'configure', muted: false}, jsep: jsep, transaction: transactionId};
```
Comment (zazabe) [inline src/webrtc/plugin/audiobridge-plugin.js]: could be streamlined with the other methods:
```js
var transactionId = ...
var transaction = ...
var message = ...
this.addTransaction(transaction);
```

File: src/webrtc/plugin/audiobridge-plugin.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+};
+
+AudiobridgePlugin.prototype.joinRoom = function(roomId) {
+  var transactionId = Transaction.generateRandomId();
+  var transaction = new Transaction(transactionId, function(response) {
+    if ('error' == response['janus'] || response['plugindata']['data']['error']) {
+      return Promise.reject(new Error('Failed room join'));
+    }
+    return Promise.resolve();
+  }.bind(this));
```
Comment (zazabe) [inline src/webrtc/plugin/audiobridge-plugin.js]: `bind()` is not needed here

**Comment by vogdb** [comment]
@zazabe please review

**Comment by zazabe** [comment]
In general, I can't really test it:
- the `AudiobridgePlugin` is not compatible with `janus-gateway-audioroom`
- either with the cm-janus url (`wss://janus-audioroom.dev.cargomedia.ch:18110`) or janus directly (`ws://janus-audioroom.dev.cargomedia.ch:18310`), i get this error: `
{code: 405, reason: "Illegal plugin to access"}` or `
{code: 460, reason: "No such plugin 'janus.plugin.audiobridge'"}`.   
  If I change the `AudiobridgePlugin.NAME`, it will failed with `{audioroom: "event", error_code: 483, error: "Missing element (id)"}`.
- the client can only communicate through a web socket, so I can't use `https://janus.conf.meetecho.com/janus/`.

I know i could setup the official janus-gateway in local, but I don't want to spend a full day on this...



- `AudiobridgePlugin ` docu missing
- we could handle more options:   
```
  audioSend: false,
  audioRecv: false,
  muted: false
```

- the `join` action used to have a `userId` property, isn't it ?  Actually, it is set as `display` in [the current client](https://github.com/cargomedia/sk/blob/master/library/VR/client-vendor/after-body-source/vr/janus/plugin/audio.js#L59), not sure about the `audiobridge`.
- `leave`, `configure`, `changeroom` are missing
- what about having a `connect` method that will toggle between `join` and `changeroom` depending if we are already connected to a room or not?
- could be useful to have a way to extend the `message`, because we will need to inject some extra data for some actions, [like `channel_data`](https://github.com/cargomedia/sk/blob/master/library/VR/client-vendor/after-body-source/vr/janus/plugin/audio.js#L58).

**Comment by zazabe** [comment]
my last 2 comments could be implemented in another PR / plugin i think...

**Comment by vogdb** [comment]
> In general, I can't really test it:

Don't worry. I can.

> - we could handle more options: ...

Audiobridge does not have such options https://janus.conf.meetecho.com/docs/janus__audiobridge_8c.html

> - the `join` action used to have a `userId` property, isn't it ?...

The problem is more general and it can be merged with your last proposition of message extending.

**Comment by vogdb** [comment]
@zazabe what do you think is better. As is now
```js
/**
 * @param {int} roomId
 * @param {Object} [options]
 * @param {int} [options.id]
 * @param....
 * @returns {Promise}
 */
AudiobridgePlugin.prototype.joinRoom = function(roomId, options) {
```
or
```js
/**
 * @param {Object} options
 * @param {int} options.roomId
 * @param {int} [options.id]
 * @param....
 * @returns {Promise}
 */
AudiobridgePlugin.prototype.joinRoom = function(options) {
```
?

**Comment by zazabe** [comment]
- about https://github.com/cargomedia/janus-gateway-js/pull/37#issuecomment-225834888, the first solution sgtm, the `roomId` is mendatory i think.
-   `audioSend` / `audioRecv` is coming from [the janus client](https://github.com/meetecho/janus-gateway/blob/master/html/janus.js#L2053)  
  - `audioRecv` is  about the `createOffer` options, see https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/WebRTC_basics#options
  - `audioSend` is about `getUserMedia` constaints 
- `mute` is more a general flag to know if the audiobridge needs to be configured with `mute: true/false`

**Comment by vogdb** [comment]
> -   `audioSend` / `audioRecv` is coming from

We don't need them. `audioSend: true` means that user shares its microphone, e.g. it does `getUserMedia({audio: true})`. `audioSend: false` => `getUserMedia({audio: false})`. So when user wants to hear someone but be muted, he mutes itself. When user wants to speak but do not hear someone, then he needs to tell the opponent to mute himself :) That is what janus.js actually does. Those options are just shorthands for this logic.

@zazabe please review

**Comment by zazabe** [comment]
> So when user wants to hear someone but be muted, he mutes itself.

Actually, it could be mendatory (and it is in our app) that a user doesn't send audio but only receive, we can't leave that choice to the user, and it's the same on the other direction.

Anyhow, let's keep it like that for this `Audiobridge` implementation, but the `Audioroom` plugin must implement such logic in the future...

Let's merge this PR!

**Comment by vogdb** [comment]
Bless you!

**Pull request merged**